Chapter 20: Multivariate Numerical Analysis

Broken down:
    
    Multivariate: Choosing from multiple variables all at once to create a model to explain one particular column (ex. which of the other variables help predict price?)
    Numerical: Working specifically with numerical variables
    Analysis: Choosing the best model with the variables that best predict a particular column

NOTE: Variable, Feature, and Column all mean the same thing!

In [1]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt

Insurance Dataset: Refer to your book for examples

AirBnb Dataset: This gives us an opportunity to use multiple variables, switch up some code, and do some actual choosing of the best model

In [3]:
df = pd.read_csv('AB_NYC_2019.csv')
df


id                                               name   host_id  \
0          2539                 Clean & quiet apt home by the park      2787   
1          2595                              Skylit Midtown Castle      2845   
2          3647                THE VILLAGE OF HARLEM....NEW YORK !      4632   
3          3831                    Cozy Entire Floor of Brownstone      4869   
4          5022   Entire Apt: Spacious Studio/Loft by central park      7192   
...         ...                                                ...       ...   
48890  36484665    Charming one bedroom - newly renovated rowhouse   8232441   
48891  36485057      Affordable room in Bushwick/East Williamsburg   6570630   
48892  36485431            Sunny Studio at Historical Neighborhood  23492952   
48893  36485609               43rd St. Time Square-cozy single bed  30985759   
48894  36487245  Trendy duplex in the very heart of Hell's Kitchen  68119814   

           host_name neighbourhood_group       neighbourhood  latitude  \
0               John            Brooklyn          Kensington  40.64749   
1           Jennifer           Manhattan             Midtown  40.75362   
2          Elisabeth           Manhattan              Harlem  40.80902   
3        LisaRoxanne            Brooklyn        Clinton Hill  40.68514   
4              Laura           Manhattan         East Harlem  40.79851   
...              ...                 ...                 ...       ...   
48890        Sabrina            Brooklyn  Bedford-Stuyvesant  40.67853   
48891        Marisol            Brooklyn            Bushwick  40.70184   
48892  Ilgar & Aysel           Manhattan              Harlem  40.81475   
48893            Taz           Manhattan      Hell's Kitchen  40.75751   
48894     Christophe           Manhattan      Hell's Kitchen  40.76404   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
0      -73.97237     Private room    149               1                  9   
1      -73.98377  Entire home/apt    225               1                 45   
2      -73.94190     Private room    150               3                  0   
3      -73.95976  Entire home/apt     89               1                270   
4      -73.94399  Entire home/apt     80              10                  9   
...          ...              ...    ...             ...                ...   
48890  -73.94995     Private room     70               2                  0   
48891  -73.93317     Private room     40               4                  0   
48892  -73.94867  Entire home/apt    115              10                  0   
48893  -73.99112      Shared room     55               1                  0   
48894  -73.98933     Private room     90               7                  0   

      last_review  reviews_per_month  calculated_host_listings_count  \
0      2018-10-19               0.21                               6   
1      2019-05-21               0.38                               2   
2             NaN                NaN                               1   
3      2019-07-05               4.64                               1   
4      2018-11-19               0.10                               1   
...           ...                ...                             ...   
48890         NaN                NaN                               2   
48891         NaN                NaN                               2   
48892         NaN                NaN                               1   
48893         NaN                NaN                               6   
48894         NaN                NaN                               1   

       availability_365  
0                   365  
1                   355  
2                   365  
3                   194  
4                     0  
...                 ...  
48890                 9  
48891                36  
48892                27  
48893                 2  
48894                23  

[48895 rows x 16 columns]

Notice: NaN means "not a number" or specifically, missing values. This will be a major problem for our analysis. Simple answer for our class: drop NaN's and RESAVE it as the new dataset

In [4]:
df = df.dropna()

LABEL: Which variable are you trying to predict? Example: Which variables predict price and how?

Plug-and-chug the variable into "label" and use this as your "y"

In [5]:
label = "price"    # this will be your response variable. What are you trying to predict? Plug-and-chug

y = df[label]
X = df.select_dtypes(np.number).assign(const=1)  #drop all categorical features and allow y-intercept to vary
X = X.drop(columns=[label])                       # X is capitalized because it represents a matrix of multiple x-columns

model = sm.OLS(y, X)
results = model.fit()

Printing the results will give a better output. This is the only way to actually see the model diagnostics.

In [8]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.038
Method:                 Least Squares   F-statistic:                     169.9
Date:                Tue, 21 Feb 2023   Prob (F-statistic):          1.80e-317
Time:                        09:01:08   Log-Likelihood:            -2.5943e+05
No. Observations:               38821   AIC:                         5.189e+05
Df Residuals:                   38811   BIC:                         5.190e+05
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
id          

PHEW! That's a lot of output. What do we do with it? 

The following diagnostics below tell us something about the model as a whole. We want to COMPARE multiple models and look for the following for the BEST FIT:

    Higher numbers: R-squared, adjust R-squared, F-Statistic, log-likelihood
    Lower numbers: p-value, AIC, BIC

What do we mean to compare multiple models? We are deciding WHICH variables to include in our model. So we START with ALL variables, and remove variables one-by-one until we get a best-fit.

    EASIEST WAY: Remove variables ONE-AT-A-TIME
    P-values: Remove the variable with the highest p-value above 0.05, UNTIL all variables are p<=0.05

KEEP the diagnostics of each of these models and compare R-Sqr, log-likelihood, p-value for F-Statistics, etc.

Minimum_nights has p=0.975, so drop it in the X.drop() function

In [29]:
label = "price"    # this will be your response variable. What are you trying to predict? Plug-and-chug

y = df[label]
X = df.select_dtypes(np.number).assign(const=1)  #drop all categorical features and allow y-intercept to vary
X = X.drop(columns=[label, "minimum_nights"])                       # X is capitalized because it represents a matrix

model = sm.OLS(y, X)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.038
Method:                 Least Squares   F-statistic:                     191.2
Date:                Mon, 20 Feb 2023   Prob (F-statistic):          1.26e-318
Time:                        14:00:56   Log-Likelihood:            -2.5943e+05
No. Observations:               38821   AIC:                         5.189e+05
Df Residuals:                   38812   BIC:                         5.190e+05
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
id          

Pay attention to any change in the diagnostics. Then drop reviews_per_month, with p=0.463

In [41]:
y = df[label]
X = df.select_dtypes(np.number).assign(const=1)  #drop all categorical features and allow y-intercept to vary
X = X.drop(columns=[label, "minimum_nights", "reviews_per_month"])                       # X is capitalized because it represents a matrix

model = sm.OLS(y, X)
results = model.fit()

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                  price   R-squared:                       0.038
Model:                            OLS   Adj. R-squared:                  0.038
Method:                 Least Squares   F-statistic:                     218.4
Date:                Mon, 20 Feb 2023   Prob (F-statistic):          1.08e-319
Time:                        15:38:45   Log-Likelihood:            -2.5943e+05
No. Observations:               38821   AIC:                         5.189e+05
Df Residuals:                   38813   BIC:                         5.189e+05
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                                     coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------
id          

You can also view individual diagnostics like aic, bic, etc., below:

In [31]:
results.aic

518880.8424216351

Looks like this last model is our best-fit. Look at all the diagnostics (p-values, etc.) and see if you can tell why this one is the best fit.

Now, in real life, we probably would drop id and host_id as well, as these are just random assigned numbers for each house. BUT if we were to look at the RAW OUTPUT above, we could write a model like this:

price = -0.00000002736id + 0.0000000034host_id + 169.87latitude - 700.7longitude - 0.1965number_of_reviews + 0.115calculated_host_listings_count - 0.0000586

const: intercept, or just a constant

    EXERCISE: Remove id and host_id. How do the diagnostics change? Does the model seem to be a better fit, even if the p-values are below 0.05?

You can check to see how well your model predicts price. 

FITTED VALUES: Take each of your rows and plug in the data for each variable, then see what the model gives you as a price
ACTUAL VALUES: These are the actual values for price for each of those rows

Compare both the fitted and actual values.

In [62]:
df['predicted'] = results.fittedvalues
df

id                                              name    host_id  \
0          2539                Clean & quiet apt home by the park       2787   
1          2595                             Skylit Midtown Castle       2845   
3          3831                   Cozy Entire Floor of Brownstone       4869   
4          5022  Entire Apt: Spacious Studio/Loft by central park       7192   
5          5099         Large Cozy 1 BR Apartment In Midtown East       7322   
...         ...                                               ...        ...   
48782  36425863        Lovely Privet Bedroom with Privet Restroom   83554966   
48790  36427429                          No.2 with queen size bed  257683179   
48799  36438336                                   Seas The Moment  211644523   
48805  36442252                     1B-1B apartment near by Metro  273841667   
48852  36455809           Cozy Private Room in Bushwick, Brooklyn   74162901   

         host_name neighbourhood_group    neighbourhood  latitude  longitude  \
0             John            Brooklyn       Kensington  40.64749  -73.97237   
1         Jennifer           Manhattan          Midtown  40.75362  -73.98377   
3      LisaRoxanne            Brooklyn     Clinton Hill  40.68514  -73.95976   
4            Laura           Manhattan      East Harlem  40.79851  -73.94399   
5            Chris           Manhattan      Murray Hill  40.74767  -73.97500   
...            ...                 ...              ...       ...        ...   
48782        Rusaa           Manhattan  Upper East Side  40.78099  -73.95366   
48790         H Ai              Queens         Flushing  40.75104  -73.81459   
48799          Ben       Staten Island      Great Kills  40.54179  -74.14275   
48805       Blaine               Bronx       Mott Haven  40.80787  -73.92400   
48852    Christine            Brooklyn         Bushwick  40.69805  -73.92801   

             room_type  price  minimum_nights  number_of_reviews last_review  \
0         Private room    149               1                  9  2018-10-19   
1      Entire home/apt    225               1                 45  2019-05-21   
3      Entire home/apt     89               1                270  2019-07-05   
4      Entire home/apt     80              10                  9  2018-11-19   
5      Entire home/apt    200               3                 74  2019-06-22   
...                ...    ...             ...                ...         ...   
48782     Private room    129               1                  1  2019-07-07   
48790     Private room     45               1                  1  2019-07-07   
48799     Private room    235               1                  1  2019-07-07   
48805  Entire home/apt    100               1                  2  2019-07-07   
48852     Private room     30               1                  1  2019-07-08   

       reviews_per_month  calculated_host_listings_count  availability_365  \
0                   0.21                               6               365   
1                   0.38                               2               355   
3                   4.64                               1               194   
4                   0.10                               1                 0   
5                   0.59                               1               129   
...                  ...                             ...               ...   
48782               1.00                               1               147   
48790               1.00                               6               339   
48799               1.00                               1                87   
48805               2.00                               1                40   
48852               1.00                               1                 1   

        predicted  
0      188.732932  
1      205.680696  
3      108.189774  
4      137.914839  
5      158.030354  
...           ...  
48782  158.723512  
48790   92.150584  
48799  245.73

Each of the following will get you the column indices

In [71]:
df.columns.get_loc('price')+1    # Because this function does not include the first column
df.columns.get_loc('predicted')+1

17

Use these indices to find the DIFFERENCE between the predicted values and the price

In [73]:
abs(df.price-df.predicted).head(10)

0     39.732932
1     19.319304
3     19.189774
4     57.914839
5     41.969646
6     59.479672
7     32.922777
8     54.335930
9      5.484981
10    11.016469
dtype: float64

In [69]:
diff = []
for row in df.itertuples(): # .itertuples() returns an indexable list of the row values
    diff.append(abs(row[17]-row[10]))
    
df['diff'] = diff
df[['price', 'predicted', 'diff']].head(10)

price   predicted       diff
0     149  188.732932  39.732932
1     225  205.680696  19.319304
3      89  108.189774  19.189774
4      80  137.914839  57.914839
5     200  158.030354  41.969646
6      60  119.479672  59.479672
7      79  111.922777  32.922777
8      79  133.335930  54.335930
9     150  155.484981   5.484981
10    135  146.016469  11.016469